### PR TITLE
Fixed #29799 -- Allowed registering lookups per field instances.

### PR DESCRIPTION
**Pull request by AllenJonathan**

### CODE REVIEW

**Comment by AllenJonathan** [comment]
This class seems to be the best way to differentiate `class` or `instance` calls (though a little less readable). I am unclear about whether tests should be written for this class separately (since this would be tested indirectly when other methods are tested).

**Comment by felixxm** [comment]
@AllenJonathan Thanks :+1: I don't think it's worth adding only `classorinstancemethod`. It's difficult to judge how useful it is without remaining elements. Can you add other pieces (even a draft) that we could discuss the implementation details?

**Comment by AllenJonathan** [comment]
Please ignore the linter errors (using a different device and so facing issues with pre-commit). `register_lookup` and `_unregister_lookup` seems to working as it should be. 

One issue that I'm facing is that all `Field` instances are wrapped in DeferredAttribute class. Which make the syntax like this: `Model.name.field.register_lookup(Lookup)` instead of: `Model.name.register_lookup(Lookup)`.

**Comment by felixxm** [comment]
> Please ignore the linter errors (using a different device and so facing issues with pre-commit). 

You don't need to use `pre-commit`, you can call `black` and `isort` on changed files.

File: tests/custom_lookups/tests.py [hunk 1]
```diff
@@ -647,3 +657,19 @@ def test_subquery_usage(self):
...
+        cls.obj1 = Company.objects.create(name='cherry', place='china')
+        cls.obj1 = Company.objects.create(name='dates', place='denmark')
+
+    def test_register_instance_lookup(self):
+        Company.name.field.register_lookup(StartsWith)
+        self.assertEqual(Company.name.field.instance_lookups, {"sw":StartsWith})
+
+    def test_unregister_instance_lookup(self):
+        Company.name.field._unregister_lookup(StartsWith)
+        self.assertEqual(Company.name.field.instance_lookups, {})
```
Comment (felixxm) [inline tests/custom_lookups/tests.py:675]: The current tests are checking only a implementation details not that it actually works. We should add a test with all three layers (class, instance, field) and check that different lookups are used.

**Review by felixxm — COMMENTED** [review]
@AllenJonathan Thanks :+1: This PR is mainly based on https://github.com/charettes/django/commit/ccc58bf2eb43722e894ca8e451cab7db1b83af43. We need much more: new context processors for registering lookups per field/instance, support reverse relations, tests with all three layers (class, instance, field) etc.

File: tests/custom_lookups/tests.py (hunk 1 above)
Comment (AllenJonathan) [reply tests/custom_lookups/tests.py:675]: Yeah. I think those working tests could be added after implementing `get_lookups` methods (would be added soon). These tests just check if the methods work as they should.

**Comment by AllenJonathan** [comment]
> @AllenJonathan Thanks 👍 This PR is mainly based on [charettes@ccc58bf](https://github.com/charettes/django/commit/ccc58bf2eb43722e894ca8e451cab7db1b83af43). We need much more: new context processors for registering lookups per field/instance, support reverse relations, tests with all three layers (class, instance, field) etc.

Any related pull request could be helpful. Also, I don't understand how context processors would come into play here, As far as I know, context_processors are used to render data in templates.

**Comment by felixxm** [comment]
> Also, I don't understand how context processors would come into play here, As far as I know, context_processors are used to render data in templates.

Context processors have many uses, I'm thinking about sth similar to the `django.test.utils.register_lookup`:
https://github.com/django/django/blob/e59d1ff5627cf9cc551e160d65858a37693de4ed/django/test/utils.py#L989-L1001

**Comment by AllenJonathan** [comment]
> > Also, I don't understand how context processors would come into play here, As far as I know, context_processors are used to render data in templates.
> 
> Context processors have many uses, I'm thinking about sth similar to the `django.test.utils.register_lookup`:
> 
> https://github.com/django/django/blob/e59d1ff5627cf9cc551e160d65858a37693de4ed/django/test/utils.py#L989-L1001

So as far as I understand, a context processor that would temporarily register (Field/instance) lookup should be added. In this way, testing would be simplified as the lookups would be unregistered automatically. So the use case would be for testing only. Am I right?

**Comment by felixxm** [comment]
> So as far as I understand, a context processor that would temporarily register (Field/instance) lookup should be added. In this way, testing would be simplified as the lookups would be unregistered automatically. So the use case would be for testing only. Am I right?

Yes.

**Comment by AllenJonathan** [comment]
1. Tests are written for both `Field` class and instances. The basic `filter` tests seem to work. Let me know about other corner cases that can added. 

> One issue that I'm facing is that all Field instances are wrapped in DeferredAttribute class. Which make the syntax like this: Model.name.field.register_lookup(Lookup) instead of: Model.name.register_lookup(Lookup).

2. This could be solved easily by adding the necessary functions in `DeferredAttribute` as class variables (functions). But I'm unsure whether to do so. Or we could also 

3. Should an error be given if an unregistered lookup is unregistered. If yes what error could it be?

4. About context processors `register_lookup` present in the `django.test.utils`. I feel no change needs to be done here. As the `register_lookup` automatically differentiates class and instances and you can see in the tests that it works perfectly fine. Let me know why exactly a new context processor should be implemented. 

I would start now on the documentation changes. Maybe reverse relations could be done after these things are completed and resolved? But some insight into what needs to be done about reverse relations would be very helpful.

File: tests/custom_lookups/tests.py [hunk 2]
```diff
@@ -647,3 +671,61 @@ def test_subquery_usage(self):
...
+            self.assertEqual(Company.place.field.get_lookup("sw"), CustomStartsWith)
+
+    def test_get_lookup_specific_instance(self):
+        with register_lookup(Company.place.field, CustomStartsWith):
+            self.assertEqual(Company.name.field.get_lookup("sw"), None)
+            self.assertEqual(Company.place.field.get_lookup("sw"), CustomStartsWith)
+
+    def test_class_lookup_with_filter(self):
+        with register_lookup(models.CharField, CustomStartsWith):
+            self.assertSequenceEqual(
```
Comment (felixxm) [inline tests/custom_lookups/tests.py:703]: Use `self.assertCountEqual()` when ordering is not specified and we have more than one expected result.

File: tests/custom_lookups/tests.py [hunk 3]
```diff
@@ -647,3 +671,61 @@ def test_subquery_usage(self):
...
+                Company.objects.filter(place__start="f"),
+                [
+                    self.obj1,
+                    self.obj3,
+                ],
+            )
+            self.assertSequenceEqual(
+                Company.objects.filter(name__start="a"),
+                [self.obj1, self.obj2, self.obj4],
+            )
```
Comment (felixxm) [inline tests/custom_lookups/tests.py:731]: Can we test this with nested context processors to see that unregistering also works? e.g.
```python
with register_lookup(
    models.CharField, CustomStartsWith, lookup_name="start")
):
    with register_lookup(Company.place.field, StartsWith4Char, lookup_name="start"):
        self.assertCountEqual(
            Company.objects.filter(place__start="f"),
            [self.obj1, self.obj3],
        )
        self.assertCountEqual(
            Company.objects.filter(name__start="a"),
            [self.obj1, self.obj2, self.obj4],
        )
    self.assertCountEqual(
        Company.objects.filter(place__start="f"),
        [self.obj1, self.obj3, self.obj4],
    )
```

**Review by felixxm — COMMENTED** [review]
@AllenJonathan Thanks :+1: 

> 1. Tests are written for both `Field` class and instances. The basic `filter` tests seem to work. Let me know about other corner cases that can added.

Some tests fail, please check them.

> > One issue that I'm facing is that all Field instances are wrapped in DeferredAttribute class. Which make the syntax like this: Model.name.field.register_lookup(Lookup) instead of: Model.name.register_lookup(Lookup).
> 
>     2. This could be solved easily by adding the necessary functions in `DeferredAttribute` as class variables (functions). But I'm unsure whether to do so. Or we could also make the syntax as `Model.name.field.register_lookup(Lookup)`. Let me know which should be done

This should be transparent to users, so proxy methods in `DeferredAttribute` sound fine to me.
 
>     3. Should an error be given if an unregistered lookup is unregistered. If yes what error could it be?

Currently we don't raise exception when unregistering an unregistered lookup, so IMO it's unnecessary. 

>     4. About context processors `register_lookup` present in the `django.test.utils`. I feel no change needs to be done here. As the `register_lookup` automatically differentiates class and instances and you can see in the tests that it works perfectly fine.

Great :+1:

> I would start now on the documentation changes. Maybe reverse relations could be done after these things are completed and resolved? But some insight into what needs to be done about reverse relations would be very helpful.

Is it not enough to rename `ForeignObject.get_lookups()` to `get_class_lookups()`?:

https://github.com/django/django/blob/f81032572107846922745b68d5b7191058fdd5f5/django/db/models/fields/related.py#L858-L864

File: tests/custom_lookups/tests.py (hunk 2 above)
Comment (AllenJonathan) [reply tests/custom_lookups/tests.py:703]: > Use `self.assertCountEqual()` when ordering is not specified and we have more than one expected result.

Sure👍

File: tests/custom_lookups/tests.py (hunk 3 above)
Comment (AllenJonathan) [reply tests/custom_lookups/tests.py]: Done👍

**Comment by AllenJonathan** [comment]
@felixxm I've made some changes. Please look into it. Let me know if there is anything else to add(or change) here. I'm going to start adding documentation changes.

File: tests/custom_lookups/tests.py [hunk 4]
```diff
@@ -647,3 +671,84 @@ def test_subquery_usage(self):
...
+                    [self.obj1, self.obj2, self.obj4],
+                )
+            self.assertCountEqual(
+                Company.objects.filter(place__start="f"),
+                [self.obj1, self.obj3, self.obj4],
+            )
+
+    def test_instance_overide_instance_lookup(self):
+        with register_lookup(Company.name, CustomStartsWith, lookup_name="start"):
+            Company.name.register_lookup(StartsWith4Char, lookup_name="start")
```
Comment (felixxm) [inline tests/custom_lookups/tests.py:751]: Should this override an instance lookup? As far as I'm aware "instance" lookups should always have a precedence as lookups with bigger granularity :thinking: (\cc @charettes).

Comment (felixxm) [inline tests/custom_lookups/tests.py:751]: Use `try ... finally` and unregister this lookup to keep tests isolated.

File: tests/custom_lookups/tests.py [hunk 5]
```diff
@@ -647,3 +671,84 @@ def test_subquery_usage(self):
...
+                )
+                self.assertCountEqual(
+                    Company.objects.filter(name__start="a"),
+                    [self.obj1, self.obj2, self.obj4],
+                )
+            self.assertCountEqual(
+                Company.objects.filter(place__start="f"),
+                [self.obj1, self.obj3, self.obj4],
+            )
+
```
Comment (felixxm) [inline tests/custom_lookups/tests.py:748]: Please add also tests for reverse relations.

File: tests/custom_lookups/tests.py (hunk 4 above)
Comment (AllenJonathan) [reply tests/custom_lookups/tests.py:751]: The registered lookups share the same `lookup_name` if any error arises the `start` lookup would be unregistered no matter what happens. I think a try-finally block is not necessary here.

Comment (charettes) [reply tests/custom_lookups/tests.py:751]: Not sure I understand, aren't these both instance lookups. Isn't that equivalent of doing

```python
Company.name.register_lookup(CustomStartsWith, lookup_name="start")
Company.name.register_lookup(StartsWith4Char, lookup_name="start")
```

And thus overrides should be working?

**Review by charettes — COMMENTED** [review]
I was wondering what are the benefits of performing any queries in `RegisterLookupTests`. 

Couldn't we simply use `get_lookup` to assert the registration works fine instead of hitting the database? It seems it would also make the tests more explicit.

File: django/db/models/query_utils.py [hunk 6]
```diff
@@ -192,20 +193,61 @@ def _check_parent_chain(self, instance):
...
+            partial_class_method = functools.partial(self.class_method, owner)
+            return self.get_partial_method(partial_class_method, self.class_method)
+        partial_instance_method = functools.partial(self.instance_method, instance)
+        return self.get_partial_method(partial_instance_method, self.class_method)
+
+    @staticmethod
+    def get_partial_method(partial_function, class_method):
+        method_name = partial_function.func.__name__
+        if method_name in ["get_class_lookups", "get_instance_lookups"]:
+            partial_function.cache_clear = class_method.cache_clear
```
Comment (charettes) [inline django/db/models/query_utils.py:227]: This seems a bit odd to have `classorinstancemethod` specialize these two here.

**Review by charettes — COMMENTED** [review]
For what it's worth I'm not entirely convinced that adding the `Model.field.register_lookup` sugar is actually worth all the boilerplate it requires.

[We already got a document pattern of to access model fields](https://docs.djangoproject.com/en/4.0/ref/models/meta/#field-access-api) so it seems a bit odd to me that we add all of this code to support an uncommon access pattern anyway.

Just my 2 cents.

File: tests/custom_lookups/tests.py (hunk 4 above)
Comment (felixxm) [reply tests/custom_lookups/tests.py:751]: Ahh, yes. I missed that both are instance lookups :facepalm:

**Comment by felixxm** [comment]
> [We already got a document pattern of to access model fields](https://docs.djangoproject.com/en/4.0/ref/models/meta/#field-access-api) so it seems a bit odd to me that we add all of this code to support an uncommon access pattern anyway.

Agreed, let's simplify this :+1:

File: django/db/models/query_utils.py (hunk 6 above)
Comment (AllenJonathan) [reply django/db/models/query_utils.py:227]: `get_class_lookups` caches lookup using `@functools.lru_cache` decorator. So I wanted `get_instance_lookups` to also have the functionality to `clear_cache`. But maybe this would be necessary as I think about it, since its only going to be used by Django developers. Maybe I'll change the test that fails
image
Replacing `get_lookups` by `get_class_lookups` would solve this.

**Comment by AllenJonathan** [comment]
> Agreed, let's simplify this 👍

Ok. I'll mention the [pattern to access model fields](https://docs.djangoproject.com/en/4.0/ref/models/meta/#field-access-api) in the `RegisterLookupMixin` docs. I too felt that simplifying the syntax using proxy methods would be too much ugly code in several places,

File: tests/custom_lookups/tests.py (hunk 4 above)
Comment (AllenJonathan) [reply tests/custom_lookups/tests.py:751]: > Not sure I understand, aren't these both instance lookups. Isn't that equivalent of doing
> 
> ```python
> Company.name.register_lookup(CustomStartsWith, lookup_name="start")
> Company.name.register_lookup(StartsWith4Char, lookup_name="start")
> ```
> 
> And thus overrides should be working?

Yeah. But wouldn't this make the test "non-isolated" if an error arises? The register_lookup context processor is used here to make sure the registered lookups are unregistered even though the test fails in between.

**Comment by AllenJonathan** [comment]
> I was wondering what are the benefits of performing any queries in `RegisterLookupTests`.
> 
> Couldn't we simply use `get_lookup` to assert the registration works fine instead of hitting the database? It seems it would also make the tests more explicit.

This could be done and it would be fine. I've been hitting the database to ensure all the code works from the user level. I think there's no harm in testing by just using `get_lookup` either. But do let me know what to do.

File: django/db/models/query_utils.py (hunk 6 above)
Comment (AllenJonathan) [reply django/db/models/query_utils.py]: Done👍

**Comment by AllenJonathan** [comment]
> For what it's worth I'm not entirely convinced that adding the `Model.field.register_lookup` sugar is actually worth all the boilerplate it requires.
> 
> [We already got a document pattern of to access model fields](https://docs.djangoproject.com/en/4.0/ref/models/meta/#field-access-api) so it seems a bit odd to me that we add all of this code to support an uncommon access pattern anyway.
> 
> Just my 2 cents.

Done👍

**Comment by felixxm** [comment]
> Couldn't we simply use get_lookup to assert the registration works fine instead of hitting the database? It seems it would also make the tests more explicit.

Agreed, let's switch to `SimpleTestCase` and `get_lookups()` assertions, without actually hitting the database.

File: docs/ref/models/lookups.txt [hunk 7]
```diff
@@ -44,30 +44,48 @@ register lookups on itself. The two prominent examples are
...
     .. method:: get_transform(transform_name)
 
-        Returns a :class:`Transform` named ``transform_name``. The default
-        implementation looks recursively on all parent classes to check if any
-        has the registered transform named ``transform_name``, returning the first
-        match.
+        Returns a :class:`Transform` named ``transform_name`` registered in the
+        class. The default implementation looks recursively on all parent classes
+        to check if any has the registered transform named ``transform_name``,
+        returning the first match.
```
Comment (felixxm) [inline docs/ref/models/lookups.txt:88]: `versionchanged` annotations are missing. We should add them to changed methods or an overall annotation in the `Registration API`, e.g.

```
.. versionchanged:: 4.2

    Support for registering lookups on instances was added.
```

File: tests/custom_lookups/models.py [hunk 8]
```diff
@@ -17,3 +17,8 @@ class Article(models.Model):
 
 class MySQLUnixTimestamp(models.Model):
     timestamp = models.PositiveIntegerField()
+
+
+class Company(models.Model):
+    name = models.CharField(max_length=20)
+    place = models.CharField(max_length=20)
```
Comment (felixxm) [inline tests/custom_lookups/models.py:24]: We can reuse `Author` model. If we need two `CharField`s then we can add e.g. `alias` to the `Author`.

File: tests/custom_lookups/tests.py [hunk 9]
```diff
@@ -316,7 +345,7 @@ def test_lookups_caching(self):
         field = Article._meta.get_field("author")
 
         # clear and re-cache
-        field.get_lookups.cache_clear()
+        field.get_class_lookups.cache_clear()
```
Comment (felixxm) [inline tests/custom_lookups/tests.py:348]: Do we need to switch to `get_class_lookups`?

File: tests/custom_lookups/tests.py [hunk 10]
```diff
@@ -647,3 +676,104 @@ def test_subquery_usage(self):
                 id__in=Author.objects.filter(age__div3=2)
             )
             self.assertSequenceEqual(qs, [a2])
+
+
+class RegisterLookupTests(TestCase):
```
Comment (felixxm) [inline tests/custom_lookups/tests.py:681]:
```suggestion
class RegisterLookupTests(SimpleTestCase):
```

**Review by felixxm — COMMENTED** [review]
@AllenJonathan Thanks for updates :+1: Please add also release notes.

File: tests/custom_lookups/tests.py (hunk 9 above)
Comment (AllenJonathan) [reply tests/custom_lookups/tests.py:348]: Here, as `field` is an instance, `get_lookups` redirects to `get_instance_lookups` doesn't have a `clear_cache` method (only `get_class_lookups` has). So clearing cache can only be done by `get_class_lookups`(which was previously called `get_lookups`).

File: tests/custom_lookups/tests.py (hunk 10 above)
Comment (AllenJonathan) [reply tests/custom_lookups/tests.py]: Done👍

File: tests/custom_lookups/models.py (hunk 8 above)
Comment (AllenJonathan) [reply tests/custom_lookups/models.py]: Done👍

File: tests/custom_lookups/tests.py (hunk 5 above)
Comment (AllenJonathan) [reply tests/custom_lookups/tests.py:704]: Done 👍

File: django/db/models/query_utils.py [hunk 11]
```diff
@@ -193,19 +193,43 @@ def _check_parent_chain(self, instance):
...
         class_lookups = [
             parent.__dict__.get("class_lookups", {}) for parent in inspect.getmro(cls)
         ]
         return cls.merge_dicts(class_lookups)
 
+    def get_instance_lookups(self):
+        class_lookups = self.get_class_lookups()
+        instance_lookups = self.__dict__.get("instance_lookups", {})
+        lookups = {**class_lookups, **instance_lookups}
+        return lookups
```
Comment (charettes) [inline django/db/models/query_utils.py:228]: Since the existence of instance lookups will be rare I suggest we avoid an unncessary dict creation when `instance_lookups` is missing

```suggestion
        if not (instance_lookups := self.__dict__.get("instance_lookups"):
            return class_lookups
        return {**class_lookups, **instance_lookups}
```

File: docs/ref/models/lookups.txt [hunk 12]
```diff
@@ -44,35 +44,57 @@ register lookups on itself. The two prominent examples are
...
 
 For a class to be a lookup, it must follow the :ref:`Query Expression API
 <query-expression>`. :class:`~Lookup` and :class:`~Transform` naturally
 follow this API.
 
+.. versionchanged:: 4.2
+
+    Support for registering lookup on Field instance was added.
```
Comment (charettes) [inline docs/ref/models/lookups.txt:96]: `Field` should be a reference to the class

**Review by charettes — COMMENTED** [review]
The feature looks pretty much complete to me at this point.

I thought there could some exploration around avoiding to deal with a cache and rely on something like a lazy `collections.ChainMap` to proxy into a generator of MRO's `class_lookups` in `get_class_lookups` but that's beyond the scope of this project and nothing points at it being a bottleneck.

It could be worth exploring as well as producing [djangobench](https://github.com/django/djangobench/) overview of the impact of this change on the time spent resolving lookups.

Does the usage of `functools.partial` and the extra function call and dict lookup have a significant impact on query creation time?

File: django/db/models/query_utils.py [hunk 13]
```diff
@@ -223,9 +223,10 @@ def get_class_lookups(cls):
 
     def get_instance_lookups(self):
         class_lookups = self.get_class_lookups()
-        instance_lookups = self.__dict__.get("instance_lookups", {})
-        lookups = {**class_lookups, **instance_lookups}
-        return lookups
+        instance_lookups = getattr(self, "instance_lookups", None)
+        if instance_lookups:
```
Comment (felixxm) [inline django/db/models/query_utils.py:227]:
```suggestion
        if instance_lookups := getattr(self, "instance_lookups", None):
```

**Comment by AllenJonathan** [comment]
I think this PR is completed. Is it ok to open this PR to merge?

**Comment by AllenJonathan** [comment]
I have opened up a [PR]() of ticket - 26511 -- Document KeyTransform and KeyTextTransform.

**Comment by felixxm** [comment]
> I think this PR is completed. Is it ok to open this PR to merge?

I will check it next week. Can you explore more and maybe produce `djangobench` overview (as suggested by Simon)?

> _"It could be worth exploring as well as producing [djangobench](https://github.com/django/djangobench/) overview of the impact of this change on the time spent resolving lookups."_
>
> _"Does the usage of functools.partial and the extra function call and dict lookup have a significant impact on query creation time?"_

**Comment by AllenJonathan** [comment]
I've run `djangobench` on my cloned local `django` repository. I've compared the `ticket_29799` branch with the `main` branch.
Here are the [results](https://paste.rs/C95). 

`qs_filter_chaining`, `query_aggregate`, `query_annotate`, `query_complex_filter`, `query_exists`, and `query_values` have shown significant difference (slower than before). `query_delete`, `query_filter` and `query_update` have significantly changes (faster than before).

File: tests/custom_lookups/tests.py [hunk 14]
```diff
@@ -647,3 +661,88 @@ def test_subquery_usage(self):
...
+        with register_lookup(article_author, RelatedMoreThan, lookup_name="mt"):
+            self.assertEqual(article_author.get_lookup("mt"), RelatedMoreThan)
+        self.assertIsNone(article_author.get_lookup("mt"))
+
+    def test_reverse_relation_lookups(self):
+        article_author = Article._meta.get_field("author")
+        with register_lookup(models.Field, CustomStartsWith, lookup_name="start"):
+            self.assertIsNone(article_author.get_lookup("start"))
+        with register_lookup(models.ForeignKey, RelatedMoreThan, lookup_name="mt"):
+            self.assertEqual(article_author.get_lookup("mt"), RelatedMoreThan)
```
Comment (felixxm) [inline tests/custom_lookups/tests.py:748]: I pushed extra tests to the https://github.com/django/django/pull/16023.

**Comment by felixxm** [comment]
I used [django-asv](https://github.com/django/django-asv) to compare this branch and the current `main` branch and there is no significant difference, see [results](https://gist.github.com/felixxm/1890ed643dcbc416eb8dd747066ca49f) :chart_with_upwards_trend: .

**Comment by felixxm** [comment]
I squashed commits and rebased after e64919ae54933ef4840f0e27e51d9fcfd55ecf4b.

**Comment by felixxm** [comment]
@AllenJonathan Thanks :+1:
